src/devices/bus/msx/cart/ram.cpp

```diff
@@ -0,0 +1,320 @@
+- Hardware Partners TM 1M
+- MK MSX Memory Mapper 1024KB
+- MSX Club Gouda Memory Mapper 1MB
+- Popolon MSX Memory Mapper
+
+2MB Memory Mapper expansions
+- 8bits4ever MEGA-MAPPER 2048KB
+- Digital KC MSX Memory Mapper 2048KB
+- MFP Extended Memory-2048K
+- MK MSX Memory Mapper 2048KB
+- MSX Club Gouda Memory Mapper 2MB
+
+4MB Memory Mapper expansions
+- 8bits4ever MEGA-MAPPER 4096KB
+- Digital KC MSX Memory Mapper 4096KB
+- MK MSX Memory Mapper 4096KB
+- MSX Club Gouda Memory Mapper 4MB
+- Padial LPE-4MB-V42KP
+- T.N.S. Addram
+- Tecnobytes Double RAM (MegaRam not supported)
+
+Not supported memory mappers:
+- 8bits4ever SD-512
+- ASCII MEM-768
+- RBSC Carnivore2
+- Classic PC 16MB Expansion Memory - 4 x 4MB memory mapper
+- GR8NET
+- MSX Cartridge Shop MegaFlashROM SCC+ SD
+- Padial LPE-4FMB-V8SKP - 4MB memory mapper + MegaRam mode
+- Playsonic - 4MB memory mapper, up to 16MB through extra register
+- Popolon Musical Memory Mapper - 1MB memory mapper + SN76489AN
+- Stichting CODE MCR-025 - 256KB memory mapper + RAM Disk + Printer buffer
+- Stichting CODE MCR-051 - 512KB memory mapper + RAM Disk + Printer buffer
+- Stichting CODE MCR-076 - 768KB memory mapper + RAM Disk + Printer buffer
+- Stichting CODE MCR-102 - 1MB memory mapper + RAM Disk + Printer buffer
+- Stichting CODE MCR-204 - 2MB memory mapper + RAM Disk + Printer buffer
+*/
+
+DECLARE_DEVICE_TYPE(MSX_CART_16K_RAM,      msx_cart_interface)
+DECLARE_DEVICE_TYPE(MSX_CART_32K_RAM,      msx_cart_interface)
+DECLARE_DEVICE_TYPE(MSX_CART_48K_RAM,      msx_cart_interface)
+DECLARE_DEVICE_TYPE(MSX_CART_64K_RAM,      msx_cart_interface)
+DECLARE_DEVICE_TYPE(MSX_CART_256K_MM_RAM,  msx_cart_interface)
+DECLARE_DEVICE_TYPE(MSX_CART_512K_MM_RAM,  msx_cart_interface)
+DECLARE_DEVICE_TYPE(MSX_CART_1024K_MM_RAM, msx_cart_interface)
+DECLARE_DEVICE_TYPE(MSX_CART_2048K_MM_RAM, msx_cart_interface)
+DECLARE_DEVICE_TYPE(MSX_CART_4096K_MM_RAM, msx_cart_interface)
+
+void msx_cart_ram_register_options(device_slot_interface &device)
+{
+	device.option_add("ram16k", MSX_CART_16K_RAM);
+	device.option_add("ram32k", MSX_CART_32K_RAM);
+	device.option_add("ram48k", MSX_CART_48K_RAM);
+	device.option_add("ram64k", MSX_CART_64K_RAM);
+	device.option_add("mm256k", MSX_CART_256K_MM_RAM);
+	device.option_add("mm512k", MSX_CART_512K_MM_RAM);
+	device.option_add("mm1024k", MSX_CART_1024K_MM_RAM);
+	device.option_add("mm2048k", MSX_CART_2048K_MM_RAM);
+	device.option_add("mm4096k", MSX_CART_4096K_MM_RAM);
+}
+
+namespace {
+
+class msx_cart_base_ram_device : public device_t, public msx_cart_interface
+{
+protected:
+	msx_cart_base_ram_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, u32 clock, u32 ram_size)
+		: device_t(mconfig, type, tag, owner, clock)
+		, msx_cart_interface(mconfig, *this)
+		, m_ram_size(ram_size)
+	{ }
+
+	virtual void device_start() override
+	{
+		m_ram.resize(m_ram_size);
+		save_item(NAME(m_ram));
+	}
+
+	std::vector<u8> m_ram;
+	u32 m_ram_size;
```

Comment:
* You don’t want to use a vector for something that doesn’t need to be resized – it’s unnecessary overhead and potential pitfalls.  Just use `std::unique_ptr<u8 []>` instead.
* Since `m_ram_size` is fixed, you should make it `const`.
* If the derived classes shouldn’t be messing with the allocation itself, you should make `m_ram` private and provide a protected accessor for obtaining the base address.

Reply:
Addressed all of the above